src/hooks/useColumnWidths.ts

```diff
@@ -40,62 +43,81 @@ export function useColumnWidths<R, SR>(
+    if (columnToAutoResize !== null) {
+      setColumnToAutoResize(null);
+      setResizedColumnWidths((resizedColumnWidths) => {
+        const oldWidth = resizedColumnWidths.get(columnToAutoResize);
+        const newWidth = measureColumnWidth(gridRef, columnToAutoResize);
+        if (newWidth !== undefined && oldWidth !== newWidth) {
+          const newResizedColumnWidths = new Map(resizedColumnWidths);
+          newResizedColumnWidths.set(columnToAutoResize, newWidth);
+          onColumnResize?.(viewportColumns.find((c) => c.key === columnToAutoResize)!, newWidth);
+          return newResizedColumnWidths;
         }
-      }
-
-      return hasChanges ? newMeasuredColumnWidths : measuredColumnWidths;
-    });
+        return resizedColumnWidths;
+      });
+    }
   }
 
   function handleColumnResize(column: CalculatedColumn<R, SR>, nextWidth: number | 'max-content') {
     const { key: resizingKey } = column;
-    const newTemplateColumns = [...templateColumns];
     const columnsToMeasure: string[] = [];
 
-    for (const { key, idx, width } of viewportColumns) {
-      if (resizingKey === key) {
-        const width = typeof nextWidth === 'number' ? `${nextWidth}px` : nextWidth;
-        newTemplateColumns[idx] = width;
-      } else if (columnsCanFlex && typeof width === 'string' && !resizedColumnWidths.has(key)) {
-        newTemplateColumns[idx] = width;
+    // remeasure all columns that can flex and are not resized by the user
+    for (const { key, width } of viewportColumns) {
+      if (
+        columnsCanFlex &&
+        resizingKey !== key &&
+        typeof width === 'string' &&
+        !resizedColumnWidths.has(key)
+      ) {
         columnsToMeasure.push(key);
       }
     }
 
-    gridRef.current!.style.gridTemplateColumns = newTemplateColumns.join(' ');
-    const measuredWidth =
-      typeof nextWidth === 'number' ? nextWidth : measureColumnWidth(gridRef, resizingKey)!;
+    if (columnsToMeasure.length > 0) {
+      setMeasuredColumnWidths((measuredColumnWidths) => {
+        const newMeasuredColumnWidths = new Map(measuredColumnWidths);
+        for (const columnKey of columnsToMeasure) {
+          newMeasuredColumnWidths.delete(columnKey);
+        }
+        return newMeasuredColumnWidths;
+      });
+    }
 
-    // TODO: remove
-    // need flushSync to keep frozen column offsets in sync
-    // we may be able to use `startTransition` or even `requestIdleCallback` instead
-    flushSync(() => {
+    if (typeof nextWidth === 'number') {
       setResizedColumnWidths((resizedColumnWidths) => {
         const newResizedColumnWidths = new Map(resizedColumnWidths);
-        newResizedColumnWidths.set(resizingKey, measuredWidth);
+        newResizedColumnWidths.set(resizingKey, nextWidth);
         return newResizedColumnWidths;
       });
-      updateMeasuredWidths(columnsToMeasure);
-    });
-
-    onColumnResize?.(column, measuredWidth);
+      onColumnResize?.(column, nextWidth);
```

Comment:
I will suggest moving the onColumnResize call to the setter function to have the same possible behavior for resizing 
https://github.com/adazzle/react-data-grid/pull/3746/files#diff-e002f181d41ce4fd6e070724347a44e4220d0332c37e5964d04cb3fc5d60d416R77
(or vice-versa)